web3/middleware/filter.py

```diff
@@ -0,0 +1,131 @@
+import itertools
+
+from eth_utils import (
+    to_hex,
+)
+from toolz.itertoolz import (
+    concat,
+)
+
+
+class RangeCounter:
+    # ie (0, 4),(5, 9),(...)
+    def __init__(self, start, stop=None, step=5):
+        self.start = start
+        self.stop = stop
+        self.step = step
+        self._count_from = itertools.count(
+            start=self.start,
+            step=self.step)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        for _start in self._count_from:
+            if self.stop and _start > self.stop:
+                raise StopIteration
+            _end = _start + (self.step - 1)
+            if self.stop and self.stop < _end:
+                _end = self.stop
+            return _start, _end
+
+
+def unload_latest_blocks(web3, fromBlock, toBlock):
+    _fromBlock = fromBlock
+
+    current_block = int(web3.eth.blockNumber)
+    if toBlock is not None and toBlock < current_block:
+        return fromBlock, toBlock
+
+    while toBlock is None or toBlock > current_block:
+        # return blocks up to current block
+        yield _fromBlock, current_block
+        _fromBlock = current_block + 1
+        current_block = int(web3.eth.blockNumber)
+        # no new blocks
+        if current_block < _fromBlock:
+            yield (None, None)
+        # reach stop_block after being passed by current_block
+        if toBlock is not None and current_block > toBlock and _fromBlock < toBlock:
+            return _fromBlock, toBlock
+
+
+def request_getLogs(
+        web3,
+        startBlock,
+        stopBlock,
+        address,
+        topics,
+        chunk_size):
+    block_chunker = RangeCounter(startBlock, stopBlock)
+    for chunk in block_chunker:
+        params = {
+            'fromBlock': chunk[0],
+            'toBlock': chunk[1],
+            'address': address,
+            'topics': topics
+        }
+        yield web3.eth.getLogs(params)
+
+
+def log_filter(
+        web3,
+        fromBlock=None,
+        toBlock=None,
+        address=None,
+        topics=None,):
+
+    if not fromBlock or fromBlock == 'latest':
```

Comment:
`not fromBlock` will result in `fromBlock == 0` not iterating from the genesis block as expected, but rather just starting at the current block which looks like a bug.